README.md
# jquery-simple-select-table

A jquery plugin for table with selectable rows and checkboxes.

## Dependencies

* jquery

## Installation

Install from npm:

    $ npm install @kanety/jquery-simple-select-table --save

## Usage

Build html as follows:

```html
<table>
  <tr>
    <th><input type="checkbox"></th>
    <th>header</th>
  </tr>
  <tr>
    <td><input type="checkbox"></td>
    <td>1</td>
  </tr>
  <tr>
    <td><input type="checkbox"></td>
    <td>2</td>
  </tr>
  <tr>
    <td><input type="checkbox"></td>
    <td>3</td>
  </tr>
</table>
```

Then run:

```javascript
$('table').simpleSelectTable();
```

### Options

Selector options:

```javascript
$('table').simpleSelectTable({
  headCheckbox: 'th :checkbox',
  dataCheckbox: 'td :checkbox',
  rowSelector: 'td'
});
```

Feature options:

```javascript
$('table').simpleSelectTable({
  useShiftClick: true,
  useCtrlClick: true,
  useUpDownKey: true
});
```

### Callbacks

```javascript
$('table').simpleSelectTable({
  ...
}).on('row:selected', function(e, $row) {
  ...
}).on('row:unselected', function(e, $row) {
  ...
});
```

## License

The library is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG

## 0.2.0

* Fix css class name.
* Bundle css with js.

## 0.1.1

* Fix selection for the current row.

## 0.1.0

* First release.


test/jquery-sipmle-select-table-spec.js
const NAMESPACE = 'simple-select-table';

describe('jquery-simple-select-table', () => {
  beforeEach(() => {
    document.body.innerHTML = __html__['index.html'];
    eval($('script').text());
  });

  describe('basic', () => {
    let $table, $tr, $td;
    let $cb, $hcb;

    beforeEach(() => {
      $table = $('#basic');
      $tr = $table.find('tr:has(td)');
      $td = $tr.find('td:last-child');
      $cb = $tr.find('td input:checkbox');
      $hcb = $table.find('th input:checkbox');
    });

    it('checks all checkboxes', () => {
      $hcb.click();
      expect($cb.filter(':checked').length).toEqual(4);
    });

    it('selects table rows', () => {
      $td.eq(0).click();
      expect($tr.eq(0).hasClass(`${NAMESPACE}-selected`)).toEqual(true);
      expect($tr.eq(1).hasClass(`${NAMESPACE}-selected`)).toEqual(false);
      expect($cb.eq(0).is(':checked')).toEqual(true);
      expect($cb.eq(1).is(':checked')).toEqual(false);
  
      $td.eq(1).click();
      expect($tr.eq(0).hasClass(`${NAMESPACE}-selected`)).toEqual(false);
      expect($tr.eq(1).hasClass(`${NAMESPACE}-selected`)).toEqual(true);
      expect($cb.eq(0).is(':checked')).toEqual(false);
      expect($cb.eq(1).is(':checked')).toEqual(true);
    });

    it('handles ctrl + click', () => {
      $td.eq(0).click();
      $td.eq(1).trigger($.Event('click', { ctrlKey: true }));
      expect($tr.eq(0).hasClass(`${NAMESPACE}-selected`)).toEqual(true);
      expect($tr.eq(1).hasClass(`${NAMESPACE}-selected`)).toEqual(false);
      expect($cb.eq(0).is(':checked')).toEqual(true);
      expect($cb.eq(1).is(':checked')).toEqual(true);
    });

    it('handles shift + click', () => {
      $td.eq(0).click();
      $td.eq(3).trigger($.Event('click', { shiftKey: true }));
      expect($tr.eq(0).hasClass(`${NAMESPACE}-selected`)).toEqual(true);
      expect($tr.eq(1).hasClass(`${NAMESPACE}-selected`)).toEqual(false);
      expect($cb.filter(':checked').length).toEqual(4);
    });

    it('handles up/down key', () => {
      $td.eq(0).click();

      $(document).trigger($.Event('keydown', { keyCode: 40 }));
      expect($tr.eq(0).hasClass(`${NAMESPACE}-selected`)).toEqual(false);
      expect($tr.eq(1).hasClass(`${NAMESPACE}-selected`)).toEqual(true);

      $(document).trigger($.Event('keydown', { keyCode: 38 }));
      expect($tr.eq(0).hasClass(`${NAMESPACE}-selected`)).toEqual(true);
      expect($tr.eq(1).hasClass(`${NAMESPACE}-selected`)).toEqual(false);
    });
  });

  describe('callbacks', () => {
    let $table, $tr, $td;
    let $message;

    beforeEach(() => {
      $table = $('#callbacks');
      $tr = $table.find('tr:has(td)');
      $td = $tr.find('td:last-child');
      $message = $('#message');
    });

    it('runs callbacks', () => {
      $td.eq(0).click();
      $td.eq(1).click();
      expect($message.text()).toContain('unselected: 1');
      expect($message.text()).toContain('selected: 2');
    });
  });
});


karma.conf.js
// Karma configuration
// Generated on Fri Nov 23 2018 10:20:11 GMT+0900 (JST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
      'node_modules/jquery/dist/jquery.min.js',
      'src/*.js',
      'test/**/*spec.js', 
      'index.html', {
        pattern: 'dist/*',
        served: true,
        included: false
      }
    ],

    proxies: {
      '/dist/': '/base/dist/',
    },


    // list of files / patterns to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'src/*.js': ['webpack', 'sourcemap'],
      'test/*spec.js': ['webpack', 'sourcemap'],
      'index.html': ['html2js']
    },

    webpack: (function() {
      var webpack = require('./webpack.config');
      webpack.mode = 'development';
      webpack.module.rules.push({
        test: /\.js$/,
        include: /src/,
        exclude: /node_modules/,
        enforce: 'post',
        use: {
          loader: 'istanbul-instrumenter-loader',
          options: { esModules: true }
        }
      });
      webpack.devtool = 'inline-source-map';
      return webpack;
    })(),


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress', 'coverage-istanbul'],

    coverageIstanbulReporter: {
      reports: ['html', 'lcovonly'],
      dir: __dirname + '/coverage'
    },

    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: (function() {
      if (process.platform === "win32") {
        return ['Chrome', 'Firefox', 'IE', 'Edge'];
      } else {
        return ['ChromeHeadless', 'FirefoxHeadless'];
      }
    })(),


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    // concurrency: Infinity
    concurrency: 1
  })
}


src/simple-select-table.js
import $ from 'jquery';

import './simple-select-table.scss';
import { NAMESPACE } from './consts';

const DEFAULTS = {
  headCheckbox: 'th :checkbox',
  dataCheckbox: 'td :checkbox',
  rowSelector: 'td:not(:has(:checkbox))',
  useShiftClick: true,
  useCtrlClick: true,
  useUpDownKey: false
};

export default class SimpleSelectTable {
  constructor(element, options = {}) {
    this.options = $.extend({}, DEFAULTS, options);

    this.$table = $(element);
    this.$document = $(document);
    this.$current = null;

    this.uid = new Date().getTime() + Math.random();
    this.namespace = `${NAMESPACE}-${this.uid}`;

    this.active = false;
    this.disableUserSelect = false;

    this.init();
  }

  init() {
    this.$table.addClass(NAMESPACE);

    this.unbind();
    this.bind();
  }

  bind() {
    if (this.options.headCheckbox) {
      this.$table.on('click.${this.namespace}', this.options.headCheckbox, (e) => {
        this.setChecked($(e.currentTarget).prop('checked'));
      });
    }

    if (this.options.rowSelector) {
      this.$table.on(`click.${this.namespace}`, this.options.rowSelector, (e) => {
        let shift = this.options.useShiftClick && e.shiftKey;
        let ctrl = this.options.useCtrlClick && e.ctrlKey;
        this.clicked($(e.currentTarget).closest('tr'), shift, ctrl);
      });
    }

    if (this.options.useShiftClick || this.options.useCtrlClick) {
      this.$table.on(`mousedown.${this.namespace}`, (e) => {
        if (e.ctrlKey || e.shiftKey) {
          this.disableUserSelect = true;
          this.$table.addClass(`${NAMESPACE}-user-select-none`);
        }
      }).on(`mouseup.${this.namespace}`, (e) => {
        if (e.ctrlKey || e.shiftKey) {
          this.disableUserSelect = false;
          this.$table.removeClass(`${NAMESPACE}-user-select-none`);
        }
      }).on(`selectstart.${this.namespace}`, (e) => {
        if (this.disableUserSelect) {
          return false;
        } else {
          return true;
        }
      });
    }

    if (this.options.useUpDownKey) {
      this.$document.on(`click.${this.namespace}`, (e) => {
        if ($.contains(this.$table[0], e.target)) {
          this.active = true;
        } else {
          this.active = false;
        }
      }).on(`keydown.${this.namespace}`, (e) => {
        if (this.active) {
          this.keydown(e.keyCode);
          e.preventDefault();
        }
      });
    }
  }

  unbind() {
    this.$table.off(`.${this.namespace}`);
    this.$document.on(`.${this.namespace}`);
  }

  checkboxes($row) {
    return this.$table.find(this.options.dataCheckbox).filter(':enabled');
  }

  checkboxAt($row) {
    return $row.find(this.options.dataCheckbox).filter(':enabled');
  }

  clicked($next, shift = false, ctrl = false) {
    let $curr = this.$current;

    if (shift && $curr && $curr.length) {
      this.setChecked(false);
      this.checkBetween($curr, $next);
    } else if (ctrl) {
      this.toggleCheck($next);
    } else {
      this.move($curr, $next);
    }
  }

  move($curr, $next) {
    this.unselect($curr);
    this.select($next);
    this.setChecked(false);
    this.check($next);
  }

  select($row) {
    if ($row && $row.length) {
      this.$current = $row;
      $row.addClass(`${NAMESPACE}-selected`);
      this.$table.trigger('row:selected', [$row]);
    }
  }

  unselect($row) {
    if ($row && $row.length) {
      $row.removeClass(`${NAMESPACE}-selected`);
      this.$table.trigger('row:unselected', [$row]);
      this.$current = null;
    }
  }

  check($row) {
    this.checkboxAt($row).prop('checked', true);
  }

  toggleCheck($row) {
    let $cb = this.checkboxAt($row);
    $cb.prop('checked', !$cb.prop('checked'));
  }

  checkBetween($row1, $row2) {
    let $rows = this.$table.find('tr');
    let n1 = $rows.index($row1);
    let n2 = $rows.index($row2);

    if (n1 > n2) {
      [n1, n2] = [n2, n1];
    }

    $rows.slice(n1, n2 + 1).each((i, row) => {
      this.checkboxAt($(row)).prop('checked', true);
    });
  }

  setChecked(checked) {
    this.checkboxes().prop('checked', checked);
  }

  keydown(keycode) {
    switch (keycode) {
    case 38:
      this.up();
      break;
    case 40:
      this.down();
      break;
    }
  }

  up() {
    let $prev = this.$current.prev(':has(td)');
    if ($prev.length) {
      this.move(this.$current, $prev);
    }
  }

  down() {
    let $next = this.$current.next(':has(td)');
    if ($next.length) {
      this.move(this.$current, $next);
    }
  }

  static getDefaults() {
    return DEFAULTS;
  }

  static setDefaults(options) {
    $.extend(DEFAULTS, options);
  }
}
